fs/fuse/fuse_dlm_cache.c

```diff
@@ -0,0 +1,520 @@
+	int lock_mode = 0;
+	pgoff_t current_start = start;
+
+	if (!cache || start > end)
+		return false;
+
+	/* Convert to lock mode if specified */
+	if (mode == FUSE_PAGE_LOCK_READ)
+		lock_mode = FUSE_PCACHE_LK_READ;
+	else if (mode == FUSE_PAGE_LOCK_WRITE)
+		lock_mode = FUSE_PCACHE_LK_WRITE;
+
+	down_read(&cache->lock);
+
+	/* Find the first range that overlaps with [start, end] */
+	range = fuse_dlm_find_overlapping(cache, start, end);
+
+	/* Check if the entire range is covered */
+	while (range && current_start <= end) {
+		/* If we're checking for a specific mode, verify it matches */
+		if (lock_mode && range->mode != lock_mode) {
+			/* Wrong lock mode */
+			up_read(&cache->lock);
+			return false;
+		}
+
+		/* Check if there's a gap before this range */
+		if (current_start < range->start) {
+			/* Found a gap */
+			up_read(&cache->lock);
+			return false;
+		}
+
+		/* Move current_start past this range */
+		current_start = range->end + 1;
+
+		/* Get next overlapping range */
+		range = fuse_page_it_iter_next(range, start, end);
+	}
+
+	/* Check if we covered the entire range */
+	if (current_start <= end) {
+		/* There's a gap at the end */
+		up_read(&cache->lock);
+		return false;
+	}
+
+	up_read(&cache->lock);
+	return true;
+}
+
+/**
+ * request a dlm lock from the fuse server
+ */
+void fuse_get_dlm_write_lock(struct file *file, loff_t offset,
+				    size_t length)
+{
+	struct fuse_file *ff = file->private_data;
+	struct inode *inode = file_inode(file);
+	struct fuse_conn *fc = get_fuse_conn(inode);
+	struct fuse_inode *fi = get_fuse_inode(inode);
+	struct fuse_mount *fm = ff->fm;
+	loff_t end = (offset + length - 1) | (PAGE_SIZE - 1);
+
+	/* note that the offset and length don't have to be page aligned here
+     * but since we only get here on writeback caching we will send out
+	 * page aligned requests */
+	offset &= PAGE_MASK;
+
+	FUSE_ARGS(args);
+	struct fuse_dlm_lock_in inarg;
+	struct fuse_dlm_lock_out outarg;
+	int err;
+
+	/* note that this can be run from different processes
+	 * at the same time. It is intentionally not protected
+	 * since a DLM implementation in the FUSE server should take care
+	 * of any races in lock requests */
+	if (fuse_dlm_range_is_locked(&fi->dlm_locked_areas, offset,
+				    end, FUSE_PAGE_LOCK_WRITE))
+		return; /* we already have this area locked */
+
+	memset(&inarg, 0, sizeof(inarg));
+	inarg.fh = ff->fh;
+
+	inarg.offset = offset;
+	inarg.size = end - offset + 1;
+	inarg.type = FUSE_DLM_LOCK_WRITE;
+
+	args.opcode = FUSE_DLM_WB_LOCK;
+	args.nodeid = get_node_id(inode);
+	args.in_numargs = 1;
+	args.in_args[0].size = sizeof(inarg);
+	args.in_args[0].value = &inarg;
+	args.out_numargs = 1;
+	args.out_args[0].size = sizeof(outarg);
+	args.out_args[0].value = &outarg;
+	err = fuse_simple_request(fm, &args);
+	if (err == -ENOSYS || outarg.locksize < length) {
+		/* fuse server told us it does not support dlm or it's seriously broken,
+		 * save the info */
+		fc->no_dlm = 1;
```

Comment:
I think `outarg.locksize < length` should abort the connection (fuse_abort_conn).